src/Featurizers/Components/NormUpdaters.h

```diff
@@ -1,298 +1,381 @@
+///  \brief         MaxNormUpdater updates max norm
+///                 max norm is the maximum absolute value of input data elements
+///
+template <typename T>
+class MaxNormUpdater {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    MaxNormUpdater(void);
+
+    void reset(void);
+
+    void update(T input);
+
+    typename TypeSelector::MaxNormTypeSelector<T>::type commit(void);
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    typename TypeSelector::MaxNormTypeSelector<T>::type _max_norm;
+
+    // flags to check if update function is called
+    // an error will be thrown if no valid input is input before commit
+    bool        _has_valid_input;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+    void update_impl(T input, std::true_type /*is_nullable*/);
+    void update_impl(T input, std::false_type /*is_nullable*/);
+
+    template <typename U>
+    void update_impl(U input);
+};
+
+} // Updaters
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  Updaters::L1NormUpdater
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Updaters::L1NormUpdater<T>::L1NormUpdater() {
+    reset();
+}
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::reset() {
+    _l1_norm = 0;
+    _has_valid_input = false;
+}
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::update(T input) {
+    update_impl(input, std::integral_constant<bool, Traits<T>::IsNullableType>());
+}
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::update_impl(T input, std::true_type /*is_nullable*/) {
+    if (Traits<T>::IsNull(input)) {
+        return;
+    }
+
+    update_impl(Traits<T>::GetNullableValue(input));
+}
+
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::update_impl(T input, std::false_type /*is_nullable*/) {
+    update_impl(input);
+}
+
+template <typename T>
+template <typename U>
+void Updaters::L1NormUpdater<T>::update_impl(U input) {
+    if (!_has_valid_input)
```

Comment:
>if (!_has_valid_input) [](start = 4, length = 22)

You can remove this if check and always set the value to true (once it is true, it will always be true).

Reply:
Good catch!